cpp-ch/local-engine/Parser/ExpressionParser.cpp

```diff
@@ -433,8 +432,31 @@ const ActionsDAG::Node * ExpressionParser::parseExpression(ActionsDAG & actions_
                         "SingularOrList options type mismatch:{} and {}",
                         elem_type->getName(),
                         option_type->getName());
+                options_type_and_field.emplace_back(type_and_field);
+            }
 
-                elem_column->insert(type_and_field.second);
+            // check tuple internal types
+            if (isTuple(elem_type) && isTuple(args[0]->result_type))
+            {
+                // align tuple inner types with nullable
+                auto tuple_type = std::static_pointer_cast<const DB::DataTypeTuple>(elem_type);
+                auto result_type = std::static_pointer_cast<const DB::DataTypeTuple>(args[0]->result_type);
+                assert(tuple_type->getElements().size() == result_type->getElements().size());
+                DataTypes new_types;
+                for (int i = 0; i < tuple_type->getElements().size(); ++i)
```

Comment:
better to refactor it with `castColumn`. You could delete dozens of codes after refactor.
1. Given the type of first argument in `in` expression
2. Given the column of second argument in `in` expression
3. Cast the column of second argument to the type of first argument.

Reply:
we can directly use the first tuple type as the type of the tuple to avoid nullable mismatch.
1. Spark guarantees that the types of tuples in the 'in' filter are completely consistent. See org.apache.spark.sql.types.DataType#equalsStructurally
2. Additionally, the mapping from Spark types to ClickHouse types is one-to-one, See TypeParser.cpp